src/fastpath/fastpath.c

```diff
@@ -8,13 +8,242 @@
 #include <fastpath/fastpath.h>
 #ifdef CONFIG_KERNEL_MCS
 #include <object/reply.h>
+#include <object/notification.h>
 #endif
 
 #ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
 #include <benchmark/benchmark_track.h>
 #endif
 #include <benchmark/benchmark_utilisation.h>
 
+#ifdef CONFIG_ARCH_ARM
+static inline
+#ifndef CONFIG_ARCH_ARM_V6
+FORCE_INLINE
+#endif
+#endif
+void NORETURN fastpath_signal(word_t cptr, word_t msgInfo)
+{
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    pde_t stored_hw_asid;
+    word_t fault_type;
+    dom_t dom;
+
+    /* Get fault type. */
+    fault_type = seL4_Fault_get_seL4_FaultType(NODE_STATE(ksCurThread)->tcbFault);
+
+    /* Check there's no saved fault */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
```

Comment:
@xurtis do we need to check for saved faults in the signal fastpath?

Reply:
I don't think so. Can you explain why we might?

Reply:
I'm not super sure but I'm thinking that maybe we shouldn't check for saved faults here so that we don't unnecessarily delay the (possibly higher priority) waiter due to a fault from the signaller. If there are saved faults for the signaller we can handle them when the signaller is next scheduled to run. That's just my guess though. What do you think?